### DISCUSSION TITLE
Add min/max temperature and humidity to `ClimateEntity` entity options

### DISCUSSION THREAD

**Discussion by gjohansson-ST** [home-assistant/architecture#1164] [Entity Models]
## Background

Currently in `ClimateEntity` there are properties for for min and max for both `temperature` and `humidity`.

These are used in the action calls for `set_temperature` and `set_humidity` to validate that the user input are within the expected range. In addition these are also used by frontend to provide the correct range in the appropriate cards.

These properties are mandatory and the developers are supposed to overwrite these with the correct min/max that is valid for the integration/device they are implementing.

The following challenges has been identified since the implementation of the validation:
- Some integrations doesn't set or identify the "correct" min/max and therefore users run into issues when they want to set temp outside the coded range, ex issue: https://github.com/home-assistant/core/issues/123612
- Some integrations has not changed from the default min/max value and therefore I doubt they are correct.
- User feature request to be able to define a range which are different than the provided by the integration. As example if the range is 10-50 the user might only want to have 30-40 to not "accidentally" set temp/humidity outside of this range.
- I don't know if the validity of [this comment](https://github.com/home-assistant/core/issues/123612#issuecomment-2407137197) but if that's correct this would help to solve that.

## Proposal

Implement entity options for min/max for temperature and humidity to allow the user to configure/change these.
All integrations should still provide a proper min/max to the extent it's possible, the properties are still mandatory from a dev implementation.

### Benefits

- The user is in control of the min/max range of the device in respect of temperature and humidity capabilities and can correct or limit the range as necessary.
- Integrations that can not properly identify the correct min/max can get "help" from the user, mainly considering generic integrations with this comment such as `zwave`, `zha`, `esphome`, etc.

### Implementation details

- Integrations that makes use of `ClimateEntity` needs to define if a user is allowed to expand the range outside of the provided min/max value (defaults to False). Mainly targets "generic" integrations as mentioned in the background. A user can always set a more narrow range within the integration provided min/max.
- Change entity min/max to read from entity options if provided by the user (validated as per above).
- Add conversion between C and F as needed (user setting is different than integration setting).
- Frontend needs to add min/max to entity options as outlined above, defaults to integration provided values (validate the user input as per above rule).
- As properties remains in place there is no need to adjust frontend cards reading other information nor current implemented validation for `set_temperature` or `set_humidity` needs to change.

### Alternatives to consider

Mainly an issue/wish for temperature settings (not seen any request for humidity) so an alternative is to leave humidity out of this.

This is an alternative to https://github.com/home-assistant/architecture/discussions/1152 which is to make min/max optional.

**Comment by emontnemery** [comment]
The proposal should explain what happens when the device has been set to a setpoint outside the range configured in Home Assistant.
Our suggestion is that a value outside the range must be allowed, we should not try to hide or suppress that. Frontend also needs to handle it.

**Reply by gjohansson-ST** [reply]
Added implementation details to outline the details (without making a PR yet).

**Reply by farmio** [reply]
Integrations might have different min/max values based on the current modes (heat / cool) or even different presets (comfort / building protection). 
Should this be considered when allowing users to change the mi /max value?

**Reply by gjohansson-ST** [reply]
Very valid point which I did not consider with this proposal. 
Need to rethink how to handle such case or even without the mode changing the integration min/max might be changing.

**Reply by emontnemery** [reply]
OK, then we put this in hold until you've had time to think the details over.

**Comment by doctorkb** [comment]
This popped up in my search for ways to override the default max humidity for my ecobee (currently, the range is 15-50%, hardcoded in HA).

Even using a "customize" entry in the YAML configuration won't override this, despite not being a limitation of ecobee's API (I can set it to higher values using custom code, but as soon as the HA component refreshes, it drops it to the max of 50).

I'm not sure who decided on the range, but it's probably something that should be available to the user -- various use cases exist, including using a hygrostat in a high-humidity environment like a greenhouse.
